ROADMAP.md

```diff
@@ -0,0 +1,74 @@
+# Rust Analyzer Roadmap 01
+
+Written on 2018-11-06, extends approximately to February 2019.
+After, we should coordinate with rustc/RLS developers to align goals and share code and experience.
+
+
+# Overall Goals
+
+The mission is:
+  * provide an excellent "code analyzed as you type" IDE experience for the Rust language,
+  * implementing the bulk of the features in Rust itself.
+
+
+High-level architecture constraints:
+  * Aim, long-term, to use the single code base with rustc.
+    It's *obvious* then the code should be shared, but OTOH, all great IDEs started as from-scratch rewrites.
+  * Don't hard-code a particular protocol or mode of operation.
+    Produce a library which could be used for implementing an LSP server, or for in-process embedding.
+  * As long as possible, stick with stable Rust (NB: we currently use beta for 2018 edition and salsa).
+
+
+# Current Goals
+
+We really should be coordinating with compiler/rls teams, but they are busy working on making Rust 2018 at the moment.
+The sync-up point will happen some time after the edition, probably early next year.
+In the mean time, the goal is to **experiment**, specifically, to figure out how a from-scratch written RLS might look like.
+
+
+## Data Storage and Protocol implementation
+
+The fundamental part of any architecture is who owned what data, how the data is mutated and how the data is exposed to user.
+For storage we use the [salsa](http://github.com/salsa-rs/salsa) library, and the model already seems solid.
+
+Most of modification is driven by the language client, but we also should support watching the file system, current implementation is a stub.
+
+**Action Item:** implement reliable file watching service.
+
+We also should extract LSP bits as a reusable library. There's already `gen_lsp_server`, but it is pretty limited.
+
+**Action Item:** try using `gen_lsp_server` in more than one language server, for example for TOML and Nix.
+
+Note that it is unclear what shape is ideal for `gen_lsp_server`.
+I'd rather avoid futures: they bring significant runtime complexity (call stacks become insane), but we don't have c10k problem to solve.
+Current interface is based on crossbeam-channel, but it's not clear if that is the best choice.
```

Comment:
Is there an open discussion around that? This might be an interesting reference for different architecture tradeoffs. (And I've been thinking of writing a rx-style thing on top of futures streams)

Reply:
Not really. Here's the problem I face with crossbeam: https://users.rust-lang.org/t/how-to-build-selectable-abstractions-on-top-of-channels/20042

Reply:
It looks like the issue has been solved in the meantime (see https://github.com/crossbeam-rs/crossbeam-channel/issues/63#issuecomment-402231157)

Reply:
Not entirely I think: with this, you'll need to use `SelectCases`es in the API instead of Channels, so the user of the api would have to use `Select` struct instead of `select!` macro. Feels heavier than the current "handle exit yourself by calling this function".